### ISSUE TITLE
Issue with `cloudpathlib` and `super()`, but sometimes works ok when in conda environment

### ISSUE THREAD

**Issue by karolzlot** [microsoft/pyright#4896]
**Describe the bug**

I'm not sure if this is bug in pyright or just incorrect typing of library.

This is the issue:

```python
Argument of type "str | CloudPath" cannot be assigned to parameter "cloud_path" of type "str | Self@S3PathWrapper" in function "__init__"
  Type "str | CloudPath" cannot be assigned to type "str | Self@S3PathWrapper"
    "CloudPath" is incompatible with "S3PathWrapper"
    Type "CloudPath" cannot be assigned to type "str | Self@S3PathWrapper"
      "CloudPath" is incompatible with "str"
      "CloudPath" is incompatible with "S3PathWrapper"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
(parameter) cloud_path: str | CloudPath
```

However it's more than it. Another thing is that this issue is always present when using python venv (requirements.txt), but it sometimes disappears when using conda env (environment.yml). It is enough that I add or remove some lines, and it will eventually happen. (check video below)

**To Reproduce**
Install venv or conda env and error is visible. Also check video to see how to reproduce additional issue with conda.

**Expected behavior**
No pyright error.

**Screenshots or Code**

```python
from typing import Union
from cloudpathlib import CloudPath, S3Path

class S3PathWrapper(S3Path):

    def __init__(
        self,
        cloud_path: Union[str, CloudPath],
    ):

        super().__init__(cloud_path=cloud_path)
```

![image](https://user-images.githubusercontent.com/11590960/229595338-7edab50a-5de4-4251-8a14-cbb7cab2802c.png)

requirements.txt
```
cloudpathlib[s3]
```

conda environment.yml
```
name: env_pyr1
channels:
  - conda-forge
  - defaults
dependencies:
  - cloudpathlib-s3

# conda env create -f environment.yml
```

Version 0.13.0 in both cases

https://user-images.githubusercontent.com/11590960/229598927-2ddd12b6-c016-4d42-8d4b-73169ac98ef8.mp4

**VS Code extension or command-line**

Tested on both:

vscode extension v2023.3.41 (Pre-Release)
vscode extension v2023.3.40

**Additional context**
I can't reproduce this thing from video anymore, it was happening reliably on vscode extension v2023.3.41 (Pre-Release), when I switched to vscode extension v2023.3.40 it disappeared and not appeared again even as I later switched again to v2023.3.41 (Pre-Release)

Heisenbug?

**Comment by erictraut** [comment]
I'm able to repro the diagnostic you're seeing. I'm not able to repro the intermittent nature though. Is it possible that you have different versions of the `cloudpathlib` installed in your different environments? The latest version (0.13.0) is a "py.typed" library, so its inline types will be honored by pyright even if `useLibraryCodeForTypes` is set to false. I don't know if older versions of the library are also "py.typed" or have the same level of type annotations. 

In any case, I'm guessing that your two environments differ in some manner that explains the behavioral difference between the two.

As for the diagnostic, pyright is correct in reporting a type violation here. The method `CloudPath.__init__` is defined as follows:

```python
DerivedCloudPath = TypeVar("DerivedCloudPath", bound="CloudPath")

class CloudPath(metaclass=CloudPathMeta):
    def __init__(
        self: DerivedCloudPath,
        cloud_path: Union[str, DerivedCloudPath],
        client: Optional["Client"] = None,
    ) -> None: ...
```

Note the use of a `TypeVar` for the `self` and `cloud_path` parameter annotations. This is an invariant type parameter. When you use a member access expression (i.e. an expression with a `.`) such as `foo.__init__`, this "binds" the method to the object it is accessed through. This supplies the `self` argument. In the case of the expression `super().__init__`, the type of the `self` argument is `S3PathWrapper`. When the method is "bound" to the `S3PathWrapper` object, it is partially specialized. That means the `DerivedCloudPath` TypeVar is replaced with `S3PathWrapper`, and the `self` parameter is removed, as follows:

```python
    def __init__(
        cloud_path: Union[str, S3PathWrapper],
        client: Optional["Client"] = None,
    ) -> None: ...
```

When you then invoke this method with the argument `cloud_path=cloud_path`, this is a type violation because `CloudPath` is not assignable to `S3PathWrapper`.

If you think that `CloudPath.__init__` should accept any `CloudPath` argument, then you should report the bug to the library maintainer. If the `CloudPath.__init__` type information is correct, then you must pass an `S3PathWrapper` object to it, not any `CloudPath`.

**Comment by karolzlot** [comment]
Both envs had 0.13.0 version. 
I can't reproduce it anymore, I will let know if I happen to see it again and be able to make it reproducible.
It's possible that for some reason adding/removing new lines was causing environment to temporarily switch to third one with old version of cloudpathlib. Still, it shouldn't happen, it was just adding / removing new lines! 

Another possibility is some vscode/pylance/pyright caching issue. As you can see in video, only when the code was in line 16 , it stopped showing diagnostic error. (two times visible on video)

-------------------------
 
I experimented and the solution for this diagnostic is to add typing to `self`.

```python
from typing import Union
from cloudpathlib import CloudPath, S3Path

class S3PathWrapper(S3Path):

    def __init__(
        self: S3Path,
        cloud_path: Union[str, S3Path],
    ):

        super().__init__(cloud_path=cloud_path)
```

It works:

![image](https://user-images.githubusercontent.com/11590960/229683109-a4ea5566-68e7-45c8-8e6d-72e7b5481170.png)


But I don't understand why it is needed in the first place. (and if it is correct thing to do)

I want to ask what I need to learn to understand it, what to google for?

I only found this: https://stackoverflow.com/questions/74111063/at-sign-in-python-type-hints-suggested-by-pylance-pyright
